### PR TITLE
Add python workflow

### PR DESCRIPTION
**Pull request by ConDai**
Checks
1.  black formatting
2.  isort formatting 
3.  types (optional)
4. linting (optional)
5. testing (optional)

Type checking, linting and testing are optional for now as they do not all pass for every python repo. They are enabled by default but can be disabled by the following flags
- `typecheckMustPass`
- `lintingMustPass`
- `testsMustPass`

Example: https://github.com/daiseeai/pyutils/pull/9/checks

### CODE REVIEW

**Comment by jordyjwilliams** [comment]
Another suggestion `pydocstyle`could be worth adding

**Review by ryan-keswick — APPROVED** [review]
Nice!

**Comment by ConDai** [comment]
> Another suggestion `pydocstyle`could be worth adding

Will add `pydocstyle`

**Comment by jordyjwilliams** [comment]
Sweet sounds good. I will review properly on Monday

**Comment by ConDai** [comment]
Bump

**Review by jordyjwilliams — CHANGES_REQUESTED** [review]
Mostly thinking that a `python_version` matrix would be cool to allow for testing/validation against multiple versions.

Would also have an open question about the `.` root use for all static checks. Happy to have a larger discussion, but if the aim of this PR is to make a reusable boiler plate etc it could be worth having these directories configurable. For example in [this draft state](https://github.com/daiseeai/api-definitions/blob/b5e1a4f0a862a4323d224173c13c92bc8f099b76/audio-processing-api/poetry_utils.py#L30-L43) when more configuration to `pystest-coverage` was desired. It could also be worth adding/uploading the coverage.


The examples here: https://github.com/marketplace/actions/install-poetry-action#testing-using-a-matrix are quite nice... I don't think the action is necessary for us here. But we could use the `caching` as we don't currently check for this. Additionally, a `code-coverage` upload could be nice.

File: .github/workflows/python.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+      - name: Set up Python 3.9
+        uses: actions/setup-python@v3
+        with:
+          python-version: 3.9
+
+      - name: Install poetry
+        run: |
+          python -m pip install --upgrade pip
+          pip install poetry
```
[the remark this answers is not in the document]
Comment (jordyjwilliams) [reply .github/workflows/python.yaml:38]: Actually as defined [here](https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#caching-packages) I would vote for modelling that flow.

File: .github/workflows/python.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+jobs:
+  build:
+    name: Build and Test
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Set up Python 3.9
+        uses: actions/setup-python@v3
```
Comment (jordyjwilliams) [inline .github/workflows/python.yaml:31]:
```suggestion
        uses: actions/setup-python@v4
```

**Comment by ryan-keswick** [comment]
The great thing about these workflows, is that the user can define what version they want to use. Based on the commit hash, so upstream changes here won't break the CICD in lots of repos. So I'd advocate for getting this out and finding what works and what doesn't

**Comment by jordyjwilliams** [comment]
ahh in that case yeah, I don't want to be a blocker here. Happy to make some suggestions in a PR at a later date. LGTM

**Review by jordyjwilliams — APPROVED** [review]

File: .github/workflows/python.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        default: true
+        type: boolean
+      lintingMustPass:
+        description: "If set to true, linting must pass"
+        default: true
+        type: boolean
+      testsMustPass:
+        description: "If set to true, testing must pass"
+        default: true
+        type: boolean
```
[the remark this answers is not in the document]
Comment (ConDai) [reply .github/workflows/python.yaml:21]: I don't follow. What's style ghaction? Also don't understand the code above? Is that a gh action? If so what's `md`?

[the remark this answers is not in the document]
Comment (jordyjwilliams) [reply .github/workflows/python.yaml:21]: nvm  it was more a consideration with why this wouldn't run on `on: [push]`. But then I realized this was a work flow definition. I meant to delete this comment.

**Comment by ConDai** [comment]
I'll create an issue so I don't forget the suggestions

**Comment by jordyjwilliams** [comment]
> I'll create an issue so I don't forget the suggestions

sgtm, I can potentially have a look at it this week and add in a quick few adjustments, but yeah for sure see benefit in getting this implemented and to use this in the current state.

**Comment by jack-davies** [comment]
We should probably keep `Request Changes` reviews for major issues that ought to block a PR until resolved, as they prevent merging-

> Submit feedback that must be addressed before merging.

**Comment by jordyjwilliams** [comment]
@jack-davies yeah my bad I guess I had used this rather informally/incorrectly in the past as more of a  hey  "x or y could be implemented or easily improved here". Sorry for my incorrect usage of `must` keyword :P

**Comment by ConDai** [comment]
[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119) :smile:

**Pull request merged**